src/garage/torch/algos/maml.py

```diff
@@ -0,0 +1,475 @@
+"""Model-Agnostic Meta-Learning (MAML) algorithm implementation for RL."""
+import collections
+import copy
+
+from dowel import tabular
+import numpy as np
+import torch
+
+from garage.misc import tensor_utils
+from garage.sampler import OnPolicyVectorizedSampler
+from garage.tf.samplers import BatchSampler
+from garage.torch.algos import _Default, make_optimizer, VPG
+from garage.torch.optimizers import DiffSGD
+from garage.torch.utils import update_module_params
+
+
+class MAML:
+    """Model-Agnostic Meta-Learning (MAML).
+
+    Args:
+        env (garage.envs.GarageEnv): A gym environment.
+        policy (garage.torch.policies.Policy): Policy.
+        baseline (garage.np.baselines.Baseline): The baseline.
+        meta_optimizer (Union[torch.optim.Optimizer, tuple]):
+            Type of optimizer.
+            This can be an optimizer type such as `torch.optim.Adam` or a tuple
+            of type and dictionary, where dictionary contains arguments to
+            initialize the optimizer e.g. `(torch.optim.Adam, {'lr' = 1e-3})`.
+        meta_batch_size (int): Number of tasks sampled per batch.
+        inner_lr (double): Adaptation learning rate.
+        num_grad_updates (int): Number of adaptation gradient steps.
+        inner_algo (garage.torch.algos.VPG): The inner algorithm used for
```

Comment:
this is really an inner loss function right? rather than reaching into an inner algorithm's private fields, can you make a module for the loss functions instead (and share it between algorithms)?

Reply:
_re:_ inner_algo is only used for its loss function.
Yes and no. The loss function of inner_algo decides how a policy is adapted to a specific task, so yes. But MAML also reuses parameters and helper functions from inner_algo, e.g. `max_path_length` and `compute_kl_constrait()`, so no. Note that, these params and helpers are not necessarily bound to inner_algo - We can also move them to MAML, although doing so will make some duplicate code.

_re:_ Modularize loss function.
This is an interesting idea. The only two available options for on-policy algorithms in garage should be the loss function of VPG and PPO, right?

Reply:
Thanks for the detailed explanation. I plan to implement MAML-TRPO, MAML-PPO and MAML-VPG by subclassing base MAML. It will be similar to how we implement tf/PPO on tf/NPO, by specifying `inner_algo` and `optimizer` in the subclass.

Reply:
Added MAML-VPG, MAML-TRPO and MAML-PPO.